File: Dynamic_Programming/3196. Maximize Total Cost of Alternating Subarrays.cpp
/*
3196. Maximize Total Cost of Alternating Subarrays
*/

//  TC:O(n) SC:O(n*2)
class Solution {
    using LL = long long;
public:
    long long maximumTotalCost(vector<int>& nums) {
        int n = nums.size();
        vector<vector<LL>> dp(n, vector<LL>(2,LLONG_MIN/2));
        dp[0][0] = nums[0];
        for(int i = 1; i < n; i++){
            if(nums[i] >= 0){
                dp[i][0] = max(dp[i-1][0] + (LL)nums[i], dp[i-1][1] + (LL)nums[i]);
            }
            else {
                dp[i][0] = max(dp[i-1][0] + (LL)nums[i], dp[i-1][1] + (LL)nums[i]);
                dp[i][1] = dp[i-1][0] - (LL)nums[i];
            }
        }
        
        return max(dp[n-1][0], dp[n-1][1]);
    }
};

/*
切割數組 求出最大subarray總和
如果是負數 就會希望跟前面組成正數可能 
子數組偶數位可以變號
負數前面有正數 就可以變正數
dp[i][j]: 走到i的時候數組的最大值可以是多少 並且前面是有變號還是沒變號 0 -> 沒變號 ,1 -> 有變號
*/

File: Dynamic_Programming/3196. Maximize Total Cost of Alternating Subarrays.py
# TC:O(n) SC:O(1)
class Solution:
    def maximumTotalCost(self, nums: List[int]) -> int:
        '''
        觀察可以得出 每個數字除了開頭元素以外都有可能會有兩種狀態
        1. 當作偶數位 * -1 該項一定要前面還有人 
        2. 當作奇數位 * 1 該項抵達以後 可以無條件切割 並且賦予前一項 偶數or奇數位
        '''

        n = len(nums)
        f0, f1 = -inf, nums[0]
        for i in range(1, n):
            prev0, prev1 = f0, f1
            f0 = nums[i] * -1 + prev1
            f1 = nums[i] + max(prev0, prev1)
        return max(f0,f1)
        '''
        遞推
        f = [[-inf] * 2 for _ in range(n)]
        f[0][1] = nums[0]
        for i in range(1, n):
            f[i][0] = nums[i] * -1 + f[i-1][1]
            f[i][1] = nums[i] + max(f[i-1][0], f[i-1][1])
        return max(f[n-1][0], f[n-1][1])
        '''
        '''
        遞歸
        @cache
        def dfs(i:int, j:int) -> int:
            if i < 0:
                return -inf if j else 0
            if j: return nums[i] + max(dfs(i-1, 0), dfs(i-1,1))
            else: return nums[i] * -1 + dfs(i-1, 1)
        return max(dfs(n-1, 0), dfs(n-1, 1))
        '''

# 劃分型DP 思路： 每一個劃分都可以切分成兩個一組 or 一個一組 
class Solution:
    def maximumTotalCost(self, nums: List[int]) -> int:
        n = len(nums)
        # 空間優化
        f0, f1 = 0, nums[0]
        for i in range(1,n):
            prev0, prev1 = f0, f1
            f0 = f1
            f1 = max(prev1 + nums[i], prev0 + nums[i-1] - nums[i])
        return f1
        '''
        遞推 
        f = [-inf] * (n+1)
        f[0] = 0
        f[1] = nums[0]
        for i in range(1,n):
            f[i+1] = max(f[i] + nums[i], f[i-1] + nums[i-1] - nums[i])
        return f[n]
        '''
        '''
        遞歸 TC:O(n) SC:O(n)
        @cache
        def dfs(i:int) ->int:
            if i == 0:
                return nums[0]
            if i < 0:
                return 0
            return max(dfs(i-1) + nums[i], dfs(i-2) + nums[i-1] - nums[i])
        return dfs(n-1)
        '''
